python-frontend/src/test/java/org/sonar/python/semantic/v2/TypeInferenceV2Test.java

```diff
@@ -581,6 +584,71 @@ def foo3(a: int = "123"):
     Assertions.assertThat(lastType.unwrappedType()).isEqualTo(INT_TYPE);
   }
 
+  @Test
+  void inferFunctionParameterTypes() {
+    FileInput root = inferTypes("""
+      def foo(param: int, *args, **kwargs):
+        ...
+      """);
+
+    var functionDef = (FunctionDef) root.statements().statements().get(0);
+    assertThat(((FunctionType) functionDef.name().typeV2()).parameters().get(0).declaredType().unwrappedType()).isEqualTo(INT_TYPE);
+    assertThat(((FunctionType) functionDef.name().typeV2()).parameters().get(1).declaredType().unwrappedType()).isEqualTo(TUPLE_TYPE);
+    assertThat(((FunctionType) functionDef.name().typeV2()).parameters().get(2).declaredType().unwrappedType()).isEqualTo(DICT_TYPE);
+  }
+
+  @Test
+  void inferFunctionParameterTypes2() {
+    FileInput root = inferTypes("""
+      class A:
+        def foo():
+          ...
+      class A:
+        def foo(param: int):
+          ...
+      def foo(param: A) -> A:
+        ...
+      """);
+
+    var functionDef = (FunctionDef) root.statements().statements().get(2);
+    assertThat(((FunctionType) functionDef.name().typeV2()).parameters().get(0).declaredType().unwrappedType()).isEqualTo(PythonType.UNKNOWN);
+    assertThat(((FunctionType) functionDef.name().typeV2()).returnType().unwrappedType()).isEqualTo(PythonType.UNKNOWN);
+  }
+
+  @Test
+  void inferFunctionParameterTypes3() {
+    FileInput root = inferTypes("""
+      class A:
+        def foo():
+          ...
+      def foo(param: A) -> A:
+        ...
+      """);
+
+    var functionDef = (FunctionDef) root.statements().statements().get(1);
+    var classType = ((ClassDef) root.statements().statements().get(0)).name().typeV2();
+
+    assertThat(((FunctionType) functionDef.name().typeV2()).parameters().get(0).declaredType().unwrappedType()).isEqualTo(classType);
+    assertThat(((FunctionType) functionDef.name().typeV2()).returnType().unwrappedType()).isEqualTo(classType);
+  }
+
+  @Test
+  void inferFunctionParameterTypes4() {
+    FileInput root = inferTypes("""
+      from re import Pattern
+      Pattern
+      def foo(param: Pattern) -> Pattern:
+        ...
+      """);
+
+    var functionDef = (FunctionDef) root.statements().statements().get(2);
+    var patternType = ((Name) ((ExpressionStatementImpl) root.statements().statements().get(1)).expressions().get(0)).typeV2();
+
+    assertThat(((FunctionType) functionDef.name().typeV2()).parameters().get(0).declaredType().unwrappedType()).isEqualTo(patternType);
+    assertThat(((FunctionType) functionDef.name().typeV2()).returnType().unwrappedType()).isEqualTo(patternType);
+  }
+
+
```

Comment:
Another test could be with type aliases, or anything else that may rely on dataflow,  such as:

```
my_alias = int
def foo(param: my_alias): ...
```

I would expect `param` not to be mapped to `int`, but since we've relied for long on the heuristic of "single assigned names", it may actually work...Having a test would be good to make sure what the behavior is.